src/main/Main.hs

```diff
@@ -848,6 +852,16 @@ execCmd ExecOpts {..} go@GlobalOpts{..} = do
                munlockFile lk -- Unlock before transferring control away.
                exec eoEnvSettings cmd args
 
+-- | Evaluate some haskell code inline.
+evalCmd :: ExecOpts -> GlobalOpts -> IO ()
+evalCmd e@ExecOpts { eoArgs = args } go@GlobalOpts {..} =
+    if length args /= 1
+    then error "Expecting ONE string argument to 'eval'"
```

Comment:
Ok, looked at the code a bit more and I don't think it'd be too hard to do this "right".  The options that would be common to `exec`/`runghc` and `eval` are already split out into a separate `ExecOptsExtra`, so you could reuse that data structure in a new `EvalArgs`.  You can factor the options parser for that out of `execOptsParser`.